### DISCUSSION TITLE
Javascript / Web Assembly / WebGL

### DISCUSSION THREAD

**Discussion by saulshanabrook** [Quansight-Labs/metadsl#134] [Ideas]
We should be able to directly compile abstract array expressions into Javascript, Web Assembly, and/or WebGL.

@mdboom gave a wonderful talk about pyiodide and showed some performance comparisons:

![](https://pbs.twimg.com/media/DsS4rXPXcAELkpA.jpg:large)

This is using (I assume) emscripten to translate the NumPy C code to webassembly and calling it. If instead, we create an abstract array expression in these benchmarks and then have a webassembly backend for uarray, we could translate the semantics of the algorithms more directly and hopefully get a speedup over translating the C code.

**Comment by mdboom** [comment]
@saulshanabrook: Your assumption is correct.  It would be really cool to see what compiling all the way to WebAssembly would give.  My hunch is that it would be a very similar factor to the difference between Numpy and that on native.
